File: Matlab/Reconstruct_by_LT.m
function Gamma_w = Reconstruct_by_LT(Rs, Ts, N, inliers, K)
    
    K_inv = inv(K);
    I = eye(3);
    skew_T = @(T)[0, -T(3,1), T(2,1); T(3,1), 0, -T(1,1); -T(2,1), T(1,1), 0];
    
    Rel_Rs = cell(1,N);
    Rel_Ts = cell(1,N);
    Rel_Cs = cell(1,N);
    Rel_Rs{1} = eye(3);
    Rel_Ts{1} = [0; 0; 0];
    for i = 2:N
        Rel_Rs{i} = Rs(:,:,i-1);
        Rel_Ts{i} = Ts(:,i-1);
    end

    %> Loop over all points
    Gamma_w = zeros(3,size(inliers, 2));
    for i = 1:size(inliers, 2)
        A = [];
        offset = 1;
        for v = 1:N
            gamma_i = K_inv * [inliers(offset:offset+1,i); 1];
            skew_gamma = skew_T(gamma_i);
            Ri = Rel_Rs{v};
            Ti = Rel_Ts{v};
            A = [A; skew_gamma*[Ri, Ti]];
            offset = offset + 2;
        end
        [~, ~, V] = svd(A);
        p3d = V(:,end);
        p3d = p3d ./ p3d(end);
        Gamma_w(:,i) = p3d(1:3,1);
    end
    
end

File: py_vo/Reconstruct_by_LT.py
import numpy as np


def Reconstruct_by_LT(Rs, Ts, N, inliers, K):
    K_inv = np.linalg.inv(K)
    skew = lambda T: np.array([[0, -T[2, 0], T[1, 0]], [T[2, 0], 0, -T[0, 0]], [-T[1, 0], T[0, 0], 0]])

    Rel_Rs = [np.eye(3) for _ in range(N)]
    Rel_Ts = [np.zeros((3, 1)) for _ in range(N)]
    for i in range(1, N):
        Rel_Rs[i] = Rs[:, :, i - 1]
        Rel_Ts[i] = Ts[:, i - 1 : i]

    Gamma_w = np.zeros((3, inliers.shape[1]))
    for i in range(inliers.shape[1]):
        A_rows = []
        offset = 0
        for v in range(N):
            gamma_i = K_inv @ np.vstack([inliers[offset : offset + 2, i : i + 1], [[1.0]]])
            offset += 2
            A_rows.append(skew(gamma_i) @ np.hstack([Rel_Rs[v], Rel_Ts[v]]))
        A = np.vstack(A_rows)
        _, _, Vt = np.linalg.svd(A)
        p3d = Vt[-1, :]
        p3d = p3d / p3d[-1]
        Gamma_w[:, i] = p3d[:3]
    return Gamma_w
